# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.cpp
# Commit: c781f30e5605
# Full Hash: c781f30e5605561c539b9d90ca625edde2e6a28b
# Author: Paul Adenot <paul@paul.cx>
# Date: 2023-12-23 09:52:03
# Regressor Bug: 1749047
# File Overlap Count: 1
# Description:
#   Bug 1749047 - Allow some decoder to take their color info from the codec string. r=chunmin
#   
#   https://searchfox.org/mozilla-central/rev/da48f565f70a57ac28862090828fbaa7fd8556f6/dom/media/VideoUtils.cpp#294
#   is one example of this.
#   
# ==============================================================================

diff -r 7cbc5704562d -r c781f30e5605 dom/media/webcodecs/VideoDecoder.cpp
--- a/dom/media/webcodecs/VideoDecoder.cpp	Fri Dec 22 21:39:15 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.cpp	Fri Dec 22 21:39:16 2023 +0000
@@ -599,12 +599,32 @@
     if (colorSpace.mMatrix.isSome()) {
       vi->mColorSpace.emplace(ToColorSpace(colorSpace.mMatrix.value()));
     }
+    // Some decoders get their primaries and transfer function from the codec
+    // string, and it's already set here. This is the case for VP9 decoders.
     if (colorSpace.mPrimaries.isSome()) {
-      vi->mColorPrimaries.emplace(ToPrimaries(colorSpace.mPrimaries.value()));
+      auto primaries = ToPrimaries(colorSpace.mPrimaries.value());
+      if (vi->mColorPrimaries.isSome()) {
+        if (vi->mColorPrimaries.value() != primaries) {
+          LOG("Conflict between decoder config and codec string, keeping codec "
+              "string primaries of %d",
+              static_cast<int>(primaries));
+        }
+      } else {
+        vi->mColorPrimaries.emplace(primaries);
+      }
     }
     if (colorSpace.mTransfer.isSome()) {
-      vi->mTransferFunction.emplace(
-          ToTransferFunction(colorSpace.mTransfer.value()));
+      auto primaries = ToTransferFunction(colorSpace.mTransfer.value());
+      if (vi->mTransferFunction.isSome()) {
+        if (vi->mTransferFunction.value() != primaries) {
+          LOG("Conflict between decoder config and codec string, keeping codec "
+              "string transfer function of %d",
+              static_cast<int>(vi->mTransferFunction.value()));
+        }
+      } else {
+        vi->mTransferFunction.emplace(
+            ToTransferFunction(colorSpace.mTransfer.value()));
+      }
     }
   }
 
